docs/design/sql/14-automatic-query-resubmission.md

```diff
@@ -0,0 +1,158 @@
+We also don't want to retry statements by default, as not all are read-only.
+
+We will analyze three scenarios:
+1. Some, but not all rows were already received when the query fails
+2. No rows were received yet, the query text starts with `SELECT` (case-insensitive, ignoring 
+       white space) 
+3. No rows were received yet, the query _doesn't_ start with `SELECT`
+
+#### Resubmission policies
+
+We propose that the user will be able to pick one of these retry policies:
+
+1. `NEVER`: the current state. If a query fails, the failure is immediately 
+   forwarded to the user
+2. `RETRY_SELECTS`: the query will be retried if:
+   - no rows were received yet
+   - the SQL text starts with `SELECT` (case-insensitive, ignoring
+     white space)
+3. `RETRY_SELECTS_ALLOW_DUPLICATES`: as before, but query can be retried
+   after it returned some rows. The already-returned rows will be returned again.
+4. `RETRY_ALL`: all queries will be retried after a failure. This option is 
+   dangerous as it will retry all queries, including DML, DDL etc. E.g. a retried
+   `CREATE MAPPING` query can fail with "object already exists", if there's no `OR
+   REPLACE` option and the first attempt actually succeeded, so the user will   
+   receive an error even though it was successful, which is very confusing.
+
+The default mode will be `NEVER` (b/w compatible).
+
+#### Ways to pick resubmission policy
+
+1. Ideally, we would be able to configure the retry policy per query, but that's
+not possible. We could add this option to Java API (to the `SqlStatement`
+class), but this is a non-standard API. We expect many clients to use JDBC or
+other standardized APIs in other languages, which don't provide this support.
+
+2. Another idea is that we could specify this as a query hint, e.g. in the form of
+`SELECT /*+ ALLOW_RETRY */`. But for this we would have to parse the query on
+the client, which we want to avoid. Another reason against this is that queries
+are often generated, such as through ORM or other SQL tools, and injecting the
+hint might not be possible or easy.
+
+3. Another option is to use a session parameter, e.g. executing `SET
+RETRY_STATEGY=<mode>`. The client will have to parse these commands, but we will
+not have to provide a full SQL parser.
+
+4. Provide a new option in `ClientConfig`, and a new JDBC URL parameter for 
+JDBC. Other standardized APIs also provide these. The disadvantage is that it
+will apply to all statements executed using that connection, that is also to DDL
+statements etc.
+
+We propose to implement options (3) and (4). Since the option (4) is feasible
+for all modes, except for not-that-useful `RETRY_ALL` mode, we consider the
+option (3) a nice-to-have feature.
+
+#### Explanation of "no rows received yet"
+
+This means that the `SqlResult.iterator()` didn't ever return from `next()`.
+However, there can be an ongoing blocking call to `next()` in another thread.
+
+#### Other configuration options
```

Comment:
What about `redoOperation` config that we have at the client? Our current invocation mechanism uses this config, but in your solution you do not even handle that config since you have a complete different retry mechanism.